token/client/src/token.rs

```diff
@@ -913,43 +913,71 @@ where
+            if let Some(transfer_hook_accounts) = &self.transfer_hook_accounts {
+                let mut instruction = instruction::transfer_checked(
+                    &self.program_id,
+                    source,
+                    &self.pubkey,
+                    destination,
+                    authority,
+                    &multisig_signers,
+                    amount,
+                    decimals,
+                )?;
+                instruction.accounts.extend(transfer_hook_accounts.clone());
+                instruction
+            } else {
+                offchain::create_transfer_checked_instruction_with_extra_metas(
+                    &self.program_id,
+                    source,
+                    self.get_address(),
+                    destination,
+                    authority,
+                    &multisig_signers,
+                    amount,
+                    decimals,
+                    fetch_account_data_fn,
+                )
+                .await
+                .map_err(|_| TokenError::AccountNotFound)?
+            }
         } else {
             #[allow(deprecated)]
-            instruction::transfer(
+            let mut instruction = instruction::transfer(
                 &self.program_id,
                 source,
                 destination,
                 authority,
                 &multisig_signers,
                 amount,
-            )?
-        };
-        if let Some(transfer_hook_accounts) = &self.transfer_hook_accounts {
-            instruction.accounts.extend(transfer_hook_accounts.clone());
-        } else {
-            #[allow(deprecated)]
-            offchain::resolve_extra_transfer_account_metas(
-                &mut instruction,
-                |address| {
-                    self.client
-                        .get_account(address)
-                        .map_ok(|opt| opt.map(|acc| acc.data))
-                },
-                self.get_address(),
-            )
-            .await
-            .map_err(|_| TokenError::AccountNotFound)?;
+            )?;
+            if let Some(transfer_hook_accounts) = &self.transfer_hook_accounts {
+                instruction.accounts.extend(transfer_hook_accounts.clone());
+            } else {
+                let mint = self.get_mint_info().await?;
+                if let Some(program_id) = transfer_hook::get_program_id(&mint) {
+                    spl_transfer_hook_interface::offchain::add_extra_account_metas_for_execute(
+                        &mut instruction,
+                        &program_id,
+                        source,
+                        self.get_address(),
+                        destination,
+                        authority,
+                        amount,
+                        fetch_account_data_fn,
+                    )
+                    .await
+                    .map_err(|_| TokenError::AccountNotFound)?;
+                }
+            };
+            instruction
```

Comment:
Maybe I'm missing something, but we can probably remove all of this -- if the decimals aren't there, then we don't do a transfer checked, which means the mint isn't passed in, so we can't use the transfer hook anyway.

Reply:
Ohhh good call, yeah we can probably do that!

Reply:
Wait, I'm not sure if this is accurate. The previous implementation was adding extra metas to either one. If a transfer is being requested, the mint should be passed in, no?

Reply:
Ahh, okay, I can update this implementation then!